core/types/subtyping.cc

```diff
@@ -38,8 +38,8 @@ TypePtr Types::any(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
 }
 
 const TypePtr underlying(const TypePtr &t1) {
-    if (auto *f = cast_type<ProxyType>(t1)) {
-        return f->underlying();
+    if (is_proxy_type(t1)) {
+        return t1.underlying();
```

Comment:
So am I correct in understanding that there's nothing checking you if you try to call `.underlying()` on a type which it does not make sense to call `underlying` on?

Can we at least make `TypePtr::underlying` have something like `ENFORCE(is_proxy_type(*this))`?

Reply:
You're looking at the wrong function. This is an existing function `underlying()`, not `TypePtr::underlying()`. `TypePtr::underlying()` throws an exception if it's not a proxy type.

Reply:
https://github.com/sorbet/sorbet/pull/3637/files#diff-8a69c6e75a3c53920d6c4a7e8b59337d5c99e35ed53d3b83b9fc0bb821726d00R53